mewarpx/mewarpx/poisson_pseudo_1d.py

```diff
@@ -0,0 +1,143 @@
+        self.nzguardrho = 2
+        self.nxguardphi = 1
+        self.nzguardphi = 1
+
+        if callable(left_voltage):
+            self.left_voltage = left_voltage
+        else:
+            self.left_voltage = lambda x: left_voltage
+
+        if callable(right_voltage):
+            self.right_voltage = right_voltage
+        else:
+            self.right_voltage = lambda x: right_voltage
+
+        self.phi = np.zeros(
+            (self.nx + 1 + 2*self.nxguardphi,
+            self.nz + 1 + 2*self.nzguardphi)
+        )
+
+        self.decompose_matrix()
+
+        print('Using direct solver.')
+        callbacks.installfieldsolver(self._run_solve)
+
+    def decompose_matrix(self):
+        """Function to build the superLU object used to solve the linear
+        system."""
+        self.nzsolve = self.nz + 1
+        self.nxsolve = self.nx + 3
+
+        # Set up the computation matrix in order to solve A*phi = rho
+        A = np.zeros(
+            (self.nzsolve*self.nxsolve, self.nzsolve*self.nxsolve)
+        )
+        kk = 0
+        for ii in range(self.nxsolve):
+            for jj in range(self.nzsolve):
+                temp = np.zeros((self.nxsolve, self.nzsolve))
+
+                if jj == 0 or jj == self.nzsolve - 1:
+                    temp[ii, jj] = 1.
+                elif jj == 1:
+                    temp[ii, jj] = -2.0
+                    temp[ii, jj-1] = 1.0
+                    temp[ii, jj+1] = 1.0
+                elif jj == self.nzsolve - 2:
+                    temp[ii, jj] = -2.0
+                    temp[ii, jj+1] = 1.0
+                    temp[ii, jj-1] = 1.0
+                elif ii == 0:
+                    temp[ii, jj] = 1.0
+                    temp[-3, jj] = -1.0
+                elif ii == self.nxsolve - 1:
+                    temp[ii, jj] = 1.0
+                    temp[2, jj] = -1.0
+                else:
+                    temp[ii, jj] = -4.0
+                    temp[ii+1, jj] = 1.0
+                    temp[ii-1, jj] = 1.0
+                    temp[ii, jj-1] = 1.0
+                    temp[ii, jj+1] = 1.0
+
+                A[kk] = temp.flatten()
+                kk += 1
+
+        A = csc_matrix(A, dtype=np.float32)
+        self.lu = sla.splu(A)
+
+    def _run_solve(self):
+        """Function run on every step to perform the required steps to solve
+        Poisson's equation."""
+        # get rho from WarpX
+        self.rho_data = mwxrun.get_gathered_rho_grid()
+        # run superLU solver only on the root processor
+        if mwxrun.me == 0:
+            self.rho_data = self.rho_data[0][:,:,0].T
+            self.solve()
+        # write phi to WarpX
+        mwxrun.set_phi_grid(self.phi.T)
+
+    def solve(self):
+        """The solution step. Includes getting the boundary potentials and
+        calculating phi from rho."""
+
+        left_voltage = self.left_voltage(mwxrun.get_t())
+        right_voltage = self.right_voltage(mwxrun.get_t())
+
+        rho = -self.rho_data[
```

Comment:
While it reads nicer to have this as a separate line, I think we should combine it with the source assignment. That avoids creating a whole new array from this division. eg
`source[1:-1,:] = (-self.dx**2 / constants.ep0) * self.rho_data[self.nxguardrho:-self.nxguardrho, self.nzguardrho:-self.nzguardrho]`

Reply:
You'll need to rewrite the nx, nz line but that's a simple arithmetic after getting self.rho_data.shape I think.

Reply:
I hear you on the efficiency of not creating a new array but I think I'll leave it like this for now since it is the same way as we've been doing it in metools (where it hasn't proved to be a bottleneck) and more importantly since I think it keeps the workflow much cleaner in terms of array shapes. If you feel strongly about it I can certainly be convinced to change it though.

Reply:
Ah, no, I didn't realize it was already split out in metools. We'll see if it's taking time when we actually benchmark it then.